src/db.c

```diff
@@ -1158,15 +1302,46 @@ void scanGenericCommand(client *c, robj *o, unsigned long cursor) {
     listRelease(keys);
 }
 
+void addSlotIdToCursor(int slot, unsigned long long int *cursor) {
+    if (!server.cluster_enabled) return;
+    /* Slot id can be -1 when iteration is over and there are no more slots to visit. */
+    if (slot < 0) return;
+    *cursor = (*cursor << CLUSTER_SLOT_MASK_BITS) | slot;
+}
+
+int getAndClearSlotIdFromCursor(unsigned long long int *cursor) {
+    if (!server.cluster_enabled) return 0;
+    int slot = (int) (*cursor & CLUSTER_SLOT_MASK);
+    *cursor = ((*cursor) & ~CLUSTER_SLOT_MASK) >> CLUSTER_SLOT_MASK_BITS;
+    return slot;
+}
+
 /* The SCAN command completely relies on scanGenericCommand. */
 void scanCommand(client *c) {
-    unsigned long cursor;
+    unsigned long long cursor;
     if (parseScanCursorOrReply(c,c->argv[1],&cursor) == C_ERR) return;
     scanGenericCommand(c,NULL,cursor);
 }
 
 void dbsizeCommand(client *c) {
-    addReplyLongLong(c,dictSize(c->db->dict));
+    redisDb *db = c->db;
+    unsigned long long int size = dbSize(db);
+    addReplyLongLong(c, size);
+}
+
+unsigned long long int dbSize(redisDb *db) {
+    return db->key_count;
+}
+
+unsigned long dbSlots(redisDb *db) {
+    unsigned long slots = 0;
+    dict *d;
+    dbIterator dbit;
+    dbIteratorInit(&dbit, db);
+    while ((d = dbIteratorNextDict(&dbit))) {
+        slots += dictSlots(d);
+    }
```

Comment:
Here, we are calculating the slots in the dictionary. Binary index tree cannot be utilized here.

Reply:
ohh, i was mixing `slots` with `keys`

Reply:
Maybe we should rename dictSlots to dictBuckets to avoid the confusion? dictSlots is only used in a handful of places within the repo. Or create an alias. It's a macro anyway.

Reply:
i don't think that's what caused my confusion, but i support the above suggestion (avoiding using the term "slots" there)

Reply:
Just +1, we talked about this offline and it's buried somewhere in one of my comments.